fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/ProtocolLogicalMetricGenerator.java

```diff
@@ -0,0 +1,177 @@
+                new LinkedHashSet<>(DefaultSystemMetricProtocols.getDefaultProtocolDictionary().values());
+        this.protocolChain = new ProtocolChain(new ArrayList<>(protocols));
+    }
+
+    @Override
+    public LinkedHashSet<LogicalMetric> bind(
+            DataApiRequestBuilder builder,
+            RequestParameters params,
+            BardConfigResources resources
+    ) {
+        return generateLogicalMetrics(
+                params.getLogicalMetrics().orElse(""),
+                resources.getMetricDictionary()
+        );
+    }
+
+    /**
+     * Validates that the bound logical metrics are valid for the table being queried.
+     *
+     * Throws {@link UnsatisfiedApiRequestConstraintsException} if logical metrics are bound before the queried logical
+     * table has been bound.
+     *
+     * @param entity  The resource constructed by the {@code bind}} method
+     * @param builder  The builder object representing the in progress DataApiRequest
+     * @param params  The request parameters sent by the client
+     * @param resources  Resources used to build the request
+     *
+     *
+     */
+    @Override
+    public void validate(
+            LinkedHashSet<LogicalMetric> entity,
+            DataApiRequestBuilder builder,
+            RequestParameters params,
+            BardConfigResources resources
+    ) {
+        if (!builder.isLogicalTableInitialized()) {
+            throw new UnsatisfiedApiRequestConstraintsException(
+                    LOGICAL_METRICS.getResourceName(),
+                    Collections.singleton(LOGICAL_TABLE.getResourceName())
+            );
+        }
+
+        if (!builder.getLogicalTableIfInitialized().isPresent()) {
+            throw new BadApiRequestException("A logical table is required for all data queries");
+        }
+
+        validateMetrics(entity, builder.getLogicalTableIfInitialized().get());
+    }
+
+    /**
+     * Extracts the list of metrics from the url metric query string and generates a set of LogicalMetrics.
+     * <p>
+     * If the query contains undefined metrics, {@link BadApiRequestException} will be
+     * thrown.
+     *
+     * This method is meant for backwards compatibility. If you do not need to use this method for that reason please
+     * prefer using a generator instance instead.
+     *
+     * @param apiMetricQuery  URL query string containing the metrics separated by ','
+     * @param metricDictionary  Metric dictionary contains the map of valid metric names and logical metric objects
+     *
+     * @return set of metric objects
+     */
+    public LinkedHashSet<LogicalMetric> generateLogicalMetrics(
+            String apiMetricQuery,
+            MetricDictionary metricDictionary
+    ) {
+        LinkedHashSet<LogicalMetric> metrics = new LinkedHashSet<>();
+        List<String> invalidMetricNames = new ArrayList<>();
+
+        List<ApiMetric> apiMetrics = protocolAntlrApiMetricParser.apply(apiMetricQuery);
```

Comment:
We have been talking about adding a hook to process data from the api request and maybe do some preprocessing of the input metrics before applying the protocols. Is this where that hook is going to go? Is there another place it is already implemented?

Reply:
by "this" I mean immediately after parsing the apiMetricQuery into ApiMetric objects.

Reply:
This is the place where request metric models can be massaged into protocol processing metric models.  Stripping out unwanted parameter values, normalizing parameter names, etc.  It's an open ended mapping chain like the apirequestmapper before it.